### PR TITLE
Expose ResponseStats via REST API

### LINKED ISSUE #2602: Auth: can we please have data from `pdns_control qtypes' via REST API?
PowerDNS authoritative reports statistics via the `servers/:id/statistics` URI (the data also shown by `pdns_control show '*'`.

Regrettably, the result from `pdns_control qtypes` is not surfaced via the API.

Would it, please, be possible to make that happen?

### PR DESCRIPTION
**Pull request by zeha**
With the intention to fix #2602

### CODE REVIEW

**Comment by zeha** [comment]
Initial stab; no tests yet.
Also I'm very unsure of the names I've picked.

Suggestions and everything welcome.

**Comment by jpmens** [comment]
```json
    {
        "name": "queries-by-qtype.A",
        "type": "StatisticItem",
        "value": "124"
    },
    {
        "name": "queries-by-qtype.NS",
        "type": "StatisticItem",
        "value": "12"
    },
    {
        "name": "queries-by-qtype.TLSA",
        "type": "StatisticItem",
        "value": "1"
    },
```

This looks good @zeha. I wonder whether the `type` ought to be different in order to differentiate it from the other values. (e.g. `QtypesItem`).

**Comment by zeha** [comment]
@jpmens other type names are a valid point, and I'll see what I can do there.

**Comment by zeha** [comment]
Review appreciated (would squash after review).

File: docs/markdown/httpapi/api_spec.md [hunk 1]
```diff
@@ -710,6 +710,13 @@ The statistic entries are dependent on the daemon type.
 Values are returned as strings.
 
 
+Authoritative: the statistics are:
+
+* with type = "StatisticItem": [Standard Performance Counters](../authoritative/performance.md#counters-variables)
+* with type = "QueryTypeStatisticItem": per-qtype query counters (same as `pdns_control qtypes`)
+* qith type = "ResponseSizeStatisticItem": response size histogram (same as `pdns_control respsizes`)
```
Comment (pieterlexis) [inline docs/markdown/httpapi/api_spec.md]: `s/^q/w/`

Comment (zeha) [inline docs/markdown/httpapi/api_spec.md]: ta, fixed

File: docs/markdown/httpapi/api_spec.md [hunk 2]
```diff
@@ -710,6 +710,13 @@ The statistic entries are dependent on the daemon type.
 Values are returned as strings.
 
 
+Authoritative: the statistics are:
+
+* with type = "StatisticItem": [Standard Performance Counters](../authoritative/performance.md#counters-variables)
+* with type = "QueryTypeStatisticItem": per-qtype query counters (same as `pdns_control qtypes`)
```
Comment (pieterlexis) [inline docs/markdown/httpapi/api_spec.md]: Awesome the pdns_control docs (in authoritative/internals.md) are missing the `qtypes`. Could you add those and link there? Or shall I fix that after this is merged?

File: pdns/ws-api.cc [hunk 3]
```diff
@@ -198,13 +201,20 @@ void apiServerStatistics(HttpRequest* req, HttpResponse* resp) {
...
+  resp_stat_t resp_size_stats;
+
+  resp_qtype_stats = g_rs.getQTypeResponseCounts();
+  resp_size_stats = g_rs.getSizeResponseCounts();
 
   Document doc;
   doc.SetArray();
-  typedef map<string, string> items_t;
-  BOOST_FOREACH(const items_t::value_type& item, items) {
+  BOOST_FOREACH(const stat_items_t::value_type& item, general_stats) {
```
Comment (pieterlexis) [inline pdns/ws-api.cc]: could you make this c++ 2011? `for(const auto& item : general_stats)`

**Comment by pieterlexis** [comment]
Code looks good afaict, and if it passes the tests you added I'm happy. @Habbie?

**Comment by pieterlexis** [comment]
```
-D_FORTIFY_SOURCE=2 -fstack-protector --param ssp-buffer-size=4 -O3   -pthread -Iext/yahttp -DHAVE_CONFIG_H -D_GNU_SOURCE -DHAVE_STRCASESTR -std=gnu++11 -DSYSCONFDIR='"/etc/powerdns/"' -DLOCALSTATEDIR='"/var/run/"'   -c -o ws-recursor.o ws-recursor.cc
make: *** No rule to make target `statbag.hh', needed by `ws-api.o'.
```

Looks liek you need to add statbag.hh to the `INCLUDES` in `build-scripts/dist-recursor`

**Comment by cmouse** [comment]
also statbag.cc if you add statbag.hh =)

**Comment by zeha** [comment]
Rebased this on top of current master.

File: docs/markdown/httpapi/api_spec.md (hunk 2 above)
Comment (zeha) [inline docs/markdown/httpapi/api_spec.md]: It'd be awesome if you could do that doc part :-)

**Comment by zeha** [comment]
Rebased once more, from my POV ok for 4.0, can add /statistics/:single_item at any later point.

**Comment by zeha** [comment]
from travis:
```
Dec 08 00:24:27 Done launching threads, ready to distribute questions
Doing DNS query to create statistic data...

; <<>> DiG 9.9.5-3ubuntu0.5-Ubuntu <<>> @127.0.0.1 -p 5300 example.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58388
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
;; QUESTION SECTION:
;example.com.			IN	A

;; AUTHORITY SECTION:
example.com.		86400	IN	SOA	ns1.example.com. ahu.example.com. 2000081501 28800 7200 604800 86400

;; Query time: 11 msec
;; SERVER: 127.0.0.1#5300(127.0.0.1)
;; WHEN: Tue Dec 08 00:24:27 UTC 2015
;; MSG SIZE  rcvd: 84

Running tests...

..Dec 08 00:24:28 HTTP Request "/api/v1/servers/localhost": Authentication by API Key failed
.SSS....F..........................Dec 08 00:24:29 Request to queue notification for domain 'test-08002924270313.org.' was processed, but no valid nameservers or ALSO-NOTIFYs found. Not notifying!
.......................SSSSSSSSS.
======================================================================
FAIL: test_read_statistics (test_Servers.Servers)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/zeha/pdns/regression-tests.api/test_Servers.py", line 46, in test_read_statistics
    self.assertIn('queries-by-qtype.A', [e['name'] for e in qtype_stats])
AssertionError: 'queries-by-qtype.A' not found in []

requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): 127.0.0.1
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/v1/servers/localhost/statistics HTTP/1.1" 200 None
```

**Comment by Habbie** [comment]
Conflicts and as you mentioned travis failure, how about a rebase?

**Comment by zeha** [comment]
Rebased, etc. To properly document this, would need OpenAPI 3.0, but I cannot get sphinxcontrib-openapi to work with it.

**Comment by Habbie** [comment]
recursor failed in Travis with lots of compile errors. I don't see how this PR could have caused them though..

**Comment by Habbie** [comment]
> recursor failed in Travis with lots of compile errors. I don't see how this PR could have caused them though..

Ah, you added a call to `make sdig` which is failing because auth has not been `./configure`d

**Comment by zeha** [comment]
> Ah, you added a call to make sdig which is failing because auth has not been ./configured

Oh, that does make sense. Should I just use `dig` instead?

**Comment by Habbie** [comment]
> Oh, that does make sense. Should I just use dig instead?

Seems fine for this use case. Also note that you have `/usr/bin/sdig` at this point, so plenty of options.

**Comment by zeha** [comment]
@Habbie (in travis) the tests use sdig from PATH now

**Pull request merged**